### PR TITLE
feat(value): introduce Array storage abstraction

### PR DESCRIPTION
**Pull request by anakrish**
Adds an opaque `Array` newtype paralleling `Object`, living under `src/value/array/` with the same module structure as `src/value/object/` (`mod.rs` / `iter.rs` / `serde.rs`).

`Array` wraps `Vec<Value>` today but exposes only a curated surface: constructors, indexed access, `first` / `last`, `push` / `pop`, `insert` / `remove`, `truncate`, `sort` / `sort_by`, `dedup`, `extend_from_slice`, `reverse`, `iter` / `iter_mut`, `cursor` (resumable), serde, conversions to `Value`, and a hand-written lexicographic `Ord`. The cursor type is re-exported behind the `rvm` feature so the follow-up `IterationState::Array` swap can land additively.

No crate-root `Array` re-export is added; the new type lives under `regorus::value::Array`, matching the storage-abstraction namespace strategy for `Object` and `Set`.

`Value::Array` is unchanged in this PR (still `Rc<Vec<Value>>`); the payload swap and call-site migration ship in the next PR.

Tests live in `src/value/tests.rs` alongside the Object tests. This also repairs pre-existing doctest examples exercised by `cargo test --doc --all-features`, so the full requested validation suite passes.

### CODE REVIEW

**Comment by anakrish** [comment]
@copilot Review this PR using all the review skills in this repo. Use a separate agent for each skill.

File: src/value/array/mod.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+    #[inline]
+    pub fn insert(&mut self, index: usize, value: Value) {
+        self.inner.insert(index, value);
+    }
+
+    #[inline]
+    pub fn remove(&mut self, index: usize) -> Value {
+        self.inner.remove(index)
+    }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/value/array/mod.rs:181]: Fixed in `05a1592`. `insert` now returns `Option<()>` (Some on success, None when `index > len`) and `remove` returns `Option<Value>` (Some with the removed element, None when out of range). No panicking path remains. Added `array_out_of_range_access_is_safe` test covering both methods plus the boundary case `insert(len, v)`.

File: src/value/array/mod.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+impl ops::Index<usize> for Array {
+    type Output = Value;
+
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        ops::Index::index(self.inner.as_slice(), index)
+    }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/value/array/mod.rs:313]: Fixed in `05a1592`. `Array::Index<usize>::index` now returns `&Value::Undefined` for out-of-range indices, matching `Value::Index<&Value>` semantics. Doc comment updated to direct callers needing to distinguish "missing" from "present-and-Undefined" to use `Array::get`. Test coverage added in `array_out_of_range_access_is_safe` including `usize::MAX`.

**Comment by anakrish** [comment]
@copilot Review this PR using the skills in the repo. Use a separate agent for each skill.

**Comment by anakrish** [comment]
Addressed the outstanding Array foundation review items:

- Updated docs/value/array.md to accurately describe the public API, rvm-gated cursor/next, and Vec-like surfaces that future backends may materialize.
- Made Array::with_capacity checked via try_reserve and return Option<Array> for oversized/unreservable capacities.
- Added memory-limit checks to Array growth paths: push/extend_from_slice now return Result; Extend/FromIterator route through the checked path.
- Added capacity boundary/overflow coverage and removed tests that asserted Value::Array with Rc<Vec<_>> directly.

Commits:
- 84d9b53 docs(value/array): accurate API surface and feature gating
- 87b26c3 feat(value/array): checked with_capacity
- ebc5165 feat(value/array): enforce memory limit on growth paths
- 25ba655 test(value/array): cover overflow + remove Vec-backend assertions

Intentionally deferred: no Value::Array migration or removal of ergonomic Vec-like APIs in this foundation PR; the docs now call out that future backends may need to materialize those views.

**Comment by anakrish** [comment]
@copilot Review this PR using all the review skills in this repo. Launch a separate agent for each skill.

**Comment by anakrish** [comment]
Addressed the follow-up review pass (Copilot consolidated findings + line comment on `pub mod array`) plus our own strict re-review.

**Memory-limit correctness**
- Dropped the process-aborting `Extend`/`FromIterator` impls; they now match the `Object` convention (delegate to inner container, no per-element cooperative-limit check). Limit-aware bulk insertion now goes through explicit fallible APIs.
- Added `Array::try_from_vec` and `Array::try_from_iter` as limit-aware counterparts to `From<Vec<Value>>` / `FromIterator`.
- `with_capacity` now also calls `check_memory_limit_if_needed` after `try_reserve` so capacity reservations cannot silently cross the cap.
- `insert` now returns `Result<()>`; bounds and limit failures both flow through `anyhow`. The limit case rolls back the inserted element.
- `push` and `extend_from_slice` roll back partial writes on limit failure so the array never observes a half-mutated state.

**API surface**
- Made `array` module private (was `pub mod`). Only `Array` (always) and `ArrayCursor` (`rvm` feature) are re-exported from `crate::value` — matching the `Object` / `ObjectCursor` pattern, no longer locking the module layout into the public API.
- Renamed the rvm-only `Cursor` to `ArrayCursor` for symmetry with `ObjectCursor` and to disambiguate at the value-module level.

**Docs + tests**
- Updated `docs/value/array.md` to reflect the private module, `ArrayCursor` rename, and the limit-enforcement contract (which methods enforce vs bypass).
- Added `try_from_vec` / `try_from_iter` happy-path tests and an out-of-bounds `insert` error test. Existing `insert` call sites migrated to the `Result` API.

Commit: `93216b2` fix(value/array): tighten memory-limit & API-surface review findings

Gates green: fmt, clippy (default + all-features + wasm bindings), 627 lib tests with `--all-features`, 242 lib tests default, 107 doc tests.

**Comment by anakrish** [comment]
@copilot Review this PR using all the review skills in this repo. Launch a separate agent for each skill.
